File: fnn/fnn.py
import numpy as np

class Layer:
    def __init__(self, input_size, output_size, activation):
        self.weights = np.random.randn(output_size, input_size)
        self.biases = np.random.randn(output_size)
        self.activation = activation

    def forward(self, x):
        z = np.dot(self.weights, x) + self.biases
        return self.activation(z)

def relu(x):
    return np.maximum(0, x)

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

class FNN:
    def __init__(self):
        self.layers = []

    def add_layer(self, input_size, output_size, activation):
        self.layers.append(Layer(input_size, output_size, activation))

    def predict(self, x):
        for layer in self.layers:
            x = layer.forward(x)
        return x

if __name__ == "__main__":
    # Example: 2 input neurons, 1 hidden layer (3 neurons, relu), 1 output neuron (sigmoid)
    model = FNN()
    model.add_layer(2, 3, relu)
    model.add_layer(3, 1, sigmoid)

    input_vec = np.array([0.5, -0.2])
    output = model.predict(input_vec)
    print("Input:", input_vec)
    print("Output:", output)

File: fnn/fnn.ts
// Simple Feedforward Neural Network (FNN) implementation in TypeScript
// Supports forward propagation with fully connected layers and ReLU/Sigmoid activations

export type Activation = 'relu' | 'sigmoid';

function relu(x: number): number {
    return Math.max(0, x);
}

function sigmoid(x: number): number {
    return 1 / (1 + Math.exp(-x));
}

function activate(x: number, activation: Activation): number {
    if (activation === 'relu') return relu(x);
    if (activation === 'sigmoid') return sigmoid(x);
    return x;
}

export class Layer {
    weights: number[][];
    biases: number[];
    activation: Activation;

    constructor(inputSize: number, outputSize: number, activation: Activation) {
        // Initialize weights and biases randomly
        this.weights = Array.from({ length: outputSize }, () =>
            Array.from({ length: inputSize }, () => Math.random() * 2 - 1)
        );
        this.biases = Array.from({ length: outputSize }, () => Math.random() * 2 - 1);
        this.activation = activation;
    }

    forward(input: number[]): number[] {
        return this.weights.map((row, i) => {
            const sum = row.reduce((acc, w, j) => acc + w * input[j], 0) + this.biases[i];
            return activate(sum, this.activation);
        });
    }
}

export class FNN {
    layers: Layer[] = [];

    addLayer(inputSize: number, outputSize: number, activation: Activation) {
        this.layers.push(new Layer(inputSize, outputSize, activation));
    }

    predict(input: number[]): number[] {
        return this.layers.reduce((inp, layer) => layer.forward(inp), input);
    }
}
